rasterio/_drivers.pyx

```diff
@@ -131,7 +131,9 @@ cdef class GDALEnv(ConfigEnv):
     def start(self):
         GDALAllRegister()
         OGRRegisterAll()
-        CPLSetErrorHandler(<CPLErrorHandler>errorHandler)
+        CPLPushErrorHandler(<CPLErrorHandler>errorHandler)
```

Comment:
Pushing and popping error handlers is, according to Even, the right way to do this.